src/smt-inout/smt_ast.cc

```diff
@@ -0,0 +1,534 @@
+  } else { // #nnn|
+    f.func_module = f.func_name.substr(1, bang_pos-1);
+  }
+} // ParseFromString
+
+std::string func_def_t::toString() const {
+  /// it also depends on whether this function has been re-written or not
+  std::string ret = "(define-fun ";
+  ret += func_name + " ";
+  // args
+  if ( args.empty() )  // not converted yet
+    ret += args_text;
+  else 
+    ret += arg_t::toString(args);
+
+  ret += " " + ret_type.toString();
+  ret += " " + func_body;
+  ret += ")";
+
+  if (! extra_comment.empty())
+    ret += " ; " + extra_comment;
+
+  return ret;
+}
+
+std::string ParseFromString(str_iterator & it, datatypes_t & dtype) {
+  it.skip();
+  it.accept("(declare-datatype ");
+  std::string mod_name;
+  if(it.head() == '|') {
+    mod_name = it.accept_current_and_read_untill("|");
+  } else {
+    mod_name = it.head_word();
+    it.skip_m(mod_name);
+  }
+  auto module_name = mod_name.substr(1,mod_name.length()-3-1); // remove | _s|
+  
+  it.skip();
+  it.accept( "((" + mod_name.substr(0,mod_name.length()-2) + "mk|" );
+  it.skip();
+  std::vector<state_var_t> & l = dtype[module_name];
+  while( it.head() == '(') {
+    auto state_var_def = state_var_t::ParseFromString(it, module_name); // it could be datatype
+    it.skip();
+    // processing
+    l.push_back(state_var_def);
+  }
+  it.accept(")))");
+  return module_name;
+}
+
+void ParseFromString(str_iterator & it, smt_file & smt) {
+  it.skip();
+  while(! it.is_end()) {
+    auto h = it.head_word();
+    if (h.find(";") == 0) {
+      std::shared_ptr<line_comment> ptr =
+        std::make_shared<line_comment>();
+      ptr->comment = it.readline_no_eol();
+      smt.items.push_back(ptr);
+      ILA_DLOG("Smt.Parse")<<"Parsed comment:"<<ptr->comment;
+
+      
+    }
+    else if (h == "(declare-datatype") {
+      smt.data_type_order.push_back(
+        ParseFromString(it,smt.datatypes) );
+    }
+    else if (h == "(define-fun") {
+      std::shared_ptr<func_def_t> ptr =
+        std::make_shared<func_def_t>();
+      func_def_t::ParseFromString(it,*ptr);
+      smt.items.push_back(ptr);
+      ILA_DLOG("Smt.Parse")<<"Parsed func:"<<ptr->func_name;
+    }
+    else {
+      ILA_ERROR << "Unrecognized:" << h <<". Stop.";
+      break;
+    }
+    it.skip();
+  }
+} // ParseFromString
+
+std::string smt_file::toString() const {
+  ILA_WARN << "SMT gen will ignore datatypes."
+    << " If not converted, this info will be lost.";
+  std::string ret;
+  for (auto && smt : items) {
+    auto line = smt->toString();
+    ret += line + 
+      ( (line.back() == '\n') ? "" : "\n" );
+  }
+  return ret;
+}
+
+
+std::string convert_to_binary(unsigned v, unsigned w) {
+  std::string ret;
+  for (unsigned i = 0; i < w; ++i) {
+    ret = std::to_string(v&1) + ret;
+    v = v >> 1;
+  }
+  ILA_ASSERT(v == 0);
+  ILA_ASSERT(ret.size() == w);
+  return "#b" + ret;
+}
+
+std::string convert_to_binary(const std::string v, unsigned radix, unsigned w) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Function parameter 'v' should be passed by const reference.](https://app.codacy.com/app/Bo-Yuan-Huang/ILAng/pullRequest?prid=4555809)